os/pm/pm_idle.c

```diff
@@ -65,12 +65,29 @@ void pm_idle(void)
 	enum pm_state_e newstate;
 #ifdef CONFIG_PM_TIMEDWAKEUP
 	clock_t delay;
+#endif
+#ifdef CONFIG_SMP
+	int cpu;
+	FAR struct tcb_s *tcb = NULL;
 #endif
 	/* Decide, which power saving level can be obtained */
 	flags = enter_critical_section();
 	newstate = pm_checkstate();
 	/* Perform state-dependent logic here */
-	pmvdbg("newstate= %d\n", newstate);
+	/* For SMP case, we need to check secondary core status
+	   If secondary core status is not in idle thread, abort
+	   the sleep and check again on next cycle
+	*/
+#ifdef CONFIG_SMP
+	if (newstate == PM_SLEEP) {
+		for (cpu = 1; cpu < CONFIG_SMP_NCPUS; cpu++) {
+			tcb = current_task(cpu);
+			if (tcb->pid != cpu) {
+				goto EXIT;
+			}
+		}
+	}
+#endif
 	/* Then force the global state change */
 	if (pm_changestate(newstate) < 0) {
```

Comment:
I think pm_changestate must not be called until cpu1 has gone to sleep. 
As I mentioned above, we need to move the cpu1 idle check into sgi4 handler. So, there might be a delay in the cpu's actually going to sleep. My suggestions is that we need to pass pm_changestate API as a function pointer to up_pm_board_sleep(). After cpu1 has gone to sleep, up_pm_board_sleep() must call the changestate API, and then put cpu0 to sleep.

Reply:
As mentioned in above comment, after we confirmed the core status, we will pause it immediately. And we will fire the SGI4 once we complete the change state to PM_SLEEP.